### PR TITLE
Preparing for 1476 (for review, not yet for merge)

### PR DESCRIPTION
**Pull request by ArneBab**
This is still missing plugin updates and the tests need to finish (already started, could not wait for them to finish before bed), so please do not pull yet.

These changes should all work from master (they are cherry-picked from next, except for one branch which started from master so I could just merge it).

Changes which went in (I still need to add these to the NEWS file):

- bertm FOAF efficiency enhancements
- bertm gif filter
- bertm ssl fixes
- nextgens logger fix
- bertm spare bitmap efficiency
- bertm reduce custom code
- xor show semi-persistent update info next to bookmarks

Plugin updates:
- Sharesite 0.4.4
- Library v37
- Freereader 6

### CODE REVIEW

**Comment by xor-freenet** [comment]
Thanks for working on this!

I've filed PRs against this branch on your repository to include my preexisting bookmark PRs which I feel are crucial to include in the release:
- https://github.com/ArneBab/fred-staging-1/pull/1
- https://github.com/ArneBab/fred-staging-1/pull/2

I have already rebased them against the current head commit to keep the required work minimal for you :)

Besides that:

It seems you did some merges as fast forward, I.e. the head commit is not a merge commit - it instead is the last commit of a branch you merged.
When merging, please always disable fast-forward so there is a merge commit:
```git merge --no-ff --gpg-sign BRANCHNAME```
This ensures that the history is more readable because it is visible which commits belonged together in a branch.
Also, it allows you to write the subject and message of the PR of the branch into the merge commit, which I would please like to ask you to always do when merging. It's important to preserve the title/message because they describe the goals/issues of the branch, which is critical to preserve for future code review in case GitHub goes down.

Further, perhaps since you cherry-picked some commits, the head commit does not have a signature and thus could be forged by an attacker - which would also be fixed by in the future merging without fast-forward and signing the merge commit. The signature also would act as proof of your review.
If you don't want to re-create the branch you can now do a workaround by pushing at least one more commit to this branch which is signed by you. This ensures that the tree you reviewed locally is proven to be identical to what is stored on GitHub.
Merging my PRs without FF as explained above would serve that purpose as long as before you merge them you validate that the current head commit hash you see locally matches what I based my PRs on.

**Comment by ArneBab** [comment]
I only did a single merge, all the rest are cherry-picking, because I did not want to deal with rebase. There is now a head commit with a signature and the tags will be signed.

**Comment by xor-freenet** [comment]
Also, WoT build 19 has been tested for 6 months, it can be considered stable.
It fixes hanging for half an hour at startup and also greatly improves captcha security.
It's on WoT's master branch, the changelog is in the tag.
If you want to include it: IIRC you have to verify the tag signature ("git verify-tag build0019" and then use e.g. "gitk" or "git describe" to ensure the head commit matches the tag), compile it, upload the binary to Freenet, and change class OfficialPlugins to link to it - and be careful to change the main version, not the testing one!
It does still use Ant, so it should be easy to compile.
Please thus do not ever use the binaries I release on the testing channel for the main release, my personal stress level currently is high enough already, I for now don't want to be responsible for binaries running on the main network.

Disclaimer: I don't know whether that is the full procedure for updating a plugin, just guessing, please also check the scripts repository for instructions.

> I only did a single merge, all the rest are cherry-picking, because I did not want to deal with rebase. There is now a head commit with a signature and the tags will be signed.

Cherry-picking with a merge commit is also possible: Before you cherry pick, switch to a new branch.

Thank you! :)

**Comment by Thynix** [comment]
Yep - see d40aa404f7c85ff3bd35886a6e1e21e952f5482b (and more generally https://github.com/freenet/fred/commits/b2556278a351da5240a91c3b03a91955a04af0ea/src/freenet/pluginmanager/OfficialPlugins.java) for reference.

**Comment by ArneBab** [comment]
@Thynix do you mean that this is one I missed in the release notes?

**Comment by ArneBab** [comment]
I added Library v37.

**Comment by nextgens** [comment]
This would be the first build in a while that would **add** code instead of removing some.

**Comment by ArneBab** [comment]
@nextgens Is that a pro (finally we get to release the stuff which has been bitrotting for half a year) or a contra (the codebase expands)? Most 80% of the net added lines of code come directly from next.

Do you see problems with the pull-request? The comment you gave in https://github.com/freenet/fred/pull/588 was addressed.

**Comment by nextgens** [comment]
It's an obvious con. You've merged only the "new visible features" part of next:

git diff --stat build01475..next
165 files changed, 2590 insertions(+), 3417 deletions(-)

yet what you propose is: +1,426 −863 (according to github)

**Comment by ArneBab** [comment]
Did you count build system changes in next the same as code changes?

You removed all the ant stuff and the gradle stuff provides ported features in fewer lines and omits some features. No wonder there are fewer lines in the build system. Also I cannot merge the jna stuff yet, but these code size savings will come in when next is ready.

So no, it’s not a con. The additional line savings in next are simply not yet ready to be released.

I merged what is release-ready. I would have preferred being able to just release next, but it’s not ready (as you also say yourself).

**Comment by nextgens** [comment]
It has always been a personal goal; the last three builds:

git diff --stat build01472..build01473
 128 files changed, 2509 insertions(+), 3116 deletions(-)

git diff --stat build01473..build01474
 11 files changed, 141 insertions(+), 44 deletions(-)

git diff --stat build01474..build01475
 101 files changed, 1985 insertions(+), 3331 deletions(-)

**Comment by Thynix** [comment]
@ArneBab Sorry, I was unclear. My intent was to reference those as demonstrations of what all is involved in upgrading a plugin.

**Comment by xor-freenet** [comment]
Bombe's review of my merged PR looked like he did review it for excessively long code as he requested to remove stuff due to being "useless noise", and I did remove the said stuff already - perhaps you didn't see those comments because my original PRs were against next so I filed new ones to re-base against 1475.

**Comment by ArneBab** [comment]
@nextgens I agree with the goal of reducing the amount of code, and I value your work on that front. A small codebase is much easier to improve than a large one, and that improvement is what I care about the most. So I’ll happily merge changes which reduce the lines of code, and I’ll prefer a functionally equivalent change which reduces the amount code to one which increases it. But I will not block changes purely because they add code.

**Comment by ArneBab** [comment]
@Thynix ah, ok. Thanks!

File: NEWS.md [hunk 1]
```diff
@@ -1,16 +1,78 @@
...
+
+master ():
+
+- FOAF efficiency enhancements for fast nodes
+- gif filter
+- ssl fixes
+- logger fix
+- spare bitmap efficiency
+- reduce custom code
+- show semi-persistent update info next to bookmarks
```
Comment (xor-freenet) [inline NEWS.md]: This isn't true: The update notifications *are* fully persistent across restart. That's the main point of one of the two PRs after all - previously when restarting you would lose *all* update notifications. Inlining the notifications into the bookmark list was the other PR.
Thus I recommend to change the wording and split up into two entries like this:
- persist notifications about updated bookmarks across restarts
- show notifications about updated bookmarks in the bookmark list itself, not only on the notifications page

Comment (ArneBab) [inline NEWS.md]: those are longer than the rest taken together. I’ll adjust the NEWS entry after the release.

Comment (xor-freenet) [inline NEWS.md]: Thanks!
I would actually prefer longer release notes in the future, I don't understand what half of the changes are about :)

**Pull request merged**